sui_storage/src/lock_service.rs

```diff
@@ -0,0 +1,552 @@
+                LockServiceCommands::Acquire {
+                    refs,
+                    tx_digest,
+                    resp,
+                } => {
+                    let res = self.acquire_locks(&refs, tx_digest);
+                    if let Err(_e) = resp.send(res) {
+                        info!("Could not respond to sender, sender dropped!");
+                    }
+                }
+                LockServiceCommands::Initialize { refs, resp } => {
+                    if let Err(_e) = resp.send(self.initialize_locks(&refs)) {
+                        info!("Could not respond to sender, sender dropped!");
+                    }
+                }
+                LockServiceCommands::RemoveLocks { refs, resp } => {
+                    if let Err(_e) = resp.send(self.delete_locks(&refs)) {
+                        info!("Could not respond to sender, sender dropped!");
+                    }
+                }
+            }
+        }
+        info!("LockService command loop stopped, the sender on other end hung up/dropped");
+    }
+
+    /// Loop to continuously process queries in a single thread
+    fn queries_loop(&self, mut receiver: Receiver<LockServiceQueries>) {
+        info!("LockService queries processing loop started");
+        while let Some(msg) = receiver.blocking_recv() {
+            match msg {
+                LockServiceQueries::GetLock { object, resp } => {
+                    if let Err(_e) = resp.send(self.get_lock(object)) {
+                        info!("Could not respond to sender!");
+                    }
+                }
+                LockServiceQueries::CheckLocksExist { objects, resp } => {
+                    if let Err(_e) = resp.send(self.locks_exist(&objects)) {
+                        info!("Could not respond to sender, sender dropped!");
+                    }
+                }
+            }
+        }
+        info!("LockService queries loop stopped, the sender on other end hung up/dropped");
+    }
+}
+
+const LOCKSERVICE_QUEUE_LEN: usize = 500;
+
+/// Atomic Sui Object locking service.
+/// Primary abstraction is an atomic op to acquire a lock on a given set of objects.
+/// Atomicity relies on single threaded loop and only one instance per authority.
+#[derive(Clone)]
+pub struct LockService {
+    sender: Sender<LockServiceCommands>,
+    query_sender: Sender<LockServiceQueries>,
+}
+
+impl LockService {
+    /// Create a new instance of LockService.  For now, the caller has to guarantee only one per data store -
+    /// namely each SuiDataStore creates its own LockService.
+    pub fn new<P: AsRef<Path>>(path: P, db_options: Option<Options>) -> Result<Self, SuiError> {
+        let inner_service = LockServiceImpl::try_open_db(path, db_options)?;
+
+        // Now, create a sync channel and spawn a thread
+        let (sender, receiver) = channel(LOCKSERVICE_QUEUE_LEN);
+        let inner2 = inner_service.clone();
+        std::thread::spawn(move || {
```

Comment:
Why would we spawn a decdicated thread here and for the query loop? Note: we run a multithreaded tokio scheduler.

Reply:
I think this is ok, since the DB does blocking IO, and is long running. If it was not a loop we could ask for a blocking task from the tokio runtime, but apparently you should not do this for long running tasks.

Reply:
I realized maybe I never responded to this, sorry @huitseeker.  I have run benchmarks and having the dedicated thread is faster than using the common Tokio scheduler.  This makes sense if you think about it, because the dedicated thread has less switching costs.  Maybe if we used a Tokio scheduler dedicated only to the lock service, but this isn't much different in practice than a dedicated thread, which is just much simpler.

Reply:
This makes sense and is generally super: would it be ok to check in a criterion bench? You'd presumably have something close to it at hand?

Reply:
Will add this as a TODO.